### ISSUE TITLE
v2 discussion

### ISSUE THREAD

**Issue by cmcaine** [cmcaine/tridactyl#434]
I'm planning a rewrite of Tridactyl's core architecture starting first week of May when my current contract ends. Please keep contributing during the rewrite - it's quite likely that your work will not be affected and if it is I will port it for you. If in doubt, ask in Riot/IRC.

The aims of this rewrite are:

1. Reduce mode related weirdness
2. Support tab local settings (tridactyl blacklist, different binds on different pages, unshadowing page binds, etc)
3. Support keyseq-style binds in all modes
4. Simplify messaging, modes and other core architecture to make future development simpler and more pleasant
    - Especially complicated stuff like completions, proper iframe support, eval, inline-editor, plugins, etc.

Method:

- Move state, parsers, etc to the content script
  - This makes tab-local state easier (mode, setl, etc)
  - Reduces messaging requirements
  - Simplifies key-processing code
- Make modes easy to write
  - Define a new and better interface
  - Standardise use of keyseq
- Rationalise UI code (hints, cmdline, completions, statusline, statusindicator, errorbar, etc) into one framework
  - Make iframe interface synchronous
  - More predictable behaviour than each module manipulating the DOM willy nilly
- Use new runtime-dispatch code instead of excmds_macros.py
  - Easier for other bits of the code to work without caring so much where it runs from
  - Macro code has odd requirements that contributors don't expect
  - Opportunity to improve and iterate on how we collect metadata from excmds, etc. for completions, documentation, validation, etc.
  - runtime-dispatch more easily extensible (hopefully, depends on --emitDecoratorMetadata)
- Stop using browser.storage event listeners for synchronising state and config
  - Ordering of these events with surrounding messaging code is undefined
  - Have to rewrite this anyway for move to content script
  - Performance issues

Desirable but not priority:

- Enable test suite development
  - Self test function + mocha shouldn't be too hard, really.

I intend to start this rewrite with `:exclaim rm -rf` and build a skeleton back up from `keyseq.ts`, `runtime-dispatch.ts` and a few bits and pieces from current tridactyl core.

Once the skeleton is in place then our existing feature code will be ported back to it. I'm not intending on rewriting all that code!

I'm hoping that this will take a couple of working weeks to do, but those working days might be spread out a bit and it may turn out to be harder than I expect. Hofstadter's law applies :)

This issue is to discuss these aims, track progress and coordinate with anyone working on related stuff (@IsaacKhor and the completions).

@bovine3dom and I will work through the new architecture on paper and post pictures here, but a bunch of the detail will be decided in implementation as we/I experiment with what's most ergonomic.

**Comment by cmcaine** [comment]
Please comment if there are aspects of current tridactyl code that you dislike or particularly like.

Or if there's anything else systematic that you think we should consider changing.

**Comment by bovine3dom** [comment]
Re 159393d and #401: key.preventDefault/stopPropagation should happen automatically (or not) if there is a relevant bind (or not).

**Comment by glacambre** [comment]
>We want to support pipelining and composition better

Maybe we should just abandon the idea of having a specific language of our own and just allow writing plain javascript which would then be eval()'ed?
For example `hint|open` would become `open(hint())`, which isn't that much more horrible and we might even get pipelining for free [thanks to ES7](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Pipeline_operator).
The reason I'm proposing this is that I really hate vimscript and implementing our own syntax sounds a lot like implementing vimscript. In my opinion, using JS would make Tridactyl closer to Emacs which I find much more beautiful architecture-wise.

>[Config and state] will be more or less completely rewritten

How do you plan on dealing with global and domain/document/tab-specific settings?

**Comment by bovine3dom** [comment]
glacambre, I think the plan is to do both. cmcaine was also slightly against us inventing a slightly worse VimL, but I think that pipelines are a bit easier to think about when you're just doing simple functions. It shouldn't be too difficult, e.g,:

```ts
export async function composite(...cmds: string[]) {
    cmds = cmds.join(" ").split("|")
    let val = ""
    for (let c of cmds) {
        val = await controller.acceptExCmd(c + " " + val)
    }
}
```

---

I think a big thing we should aim for in v2 is better documentation for our config. We could get some of this for free with TypeDoc if we changed configuration to be objects or maps rather than our own hacky `o({})` things, and then we can give them docstrings too.

**Comment by glacambre** [comment]
Something similar to pipes could be implemented in various ways. Here's what I'm thinking about:
```js
function parenpipe(e1) {
    return (e2) => { return parenpipe(e2(e1)) }
}
function comapipe(...args) {
    return args.reduce((prev, cur) => { return cur(prev) })
} 
```
A user who wants to use piping to send a link to mpv could do this:
```js
parenpipe(hint())(link => exclaim("mpv " + link))(do_something_with_mpv_output)
comapipe(hint(), link => exclaim("mpv " + link), do_something_with_mpv_output)
```
If we gave one of these implementations a shorter name (e.g. `p`), I think piping could be very readable and easy to use. And of course, it would also reduce the amount of code that needs to be maintained.

**Comment by bovine3dom** [comment]
Regarding
> Easier for other bits of the code to work without caring so much where it runs from

We should ensure that it's easy for `js` to call things that live in the background without it being too difficult; it's confusing that it "just works" for most stuff in `jsb`.

**Comment by tjim** [comment]
Regarding using javascript instead of inventing a vimscript, my vote is for javascript. I'm coming from conkeror (unfortunately, dying along with XUL) where that's a great feature. Just want to point out that you can experiment with this already in tridactyl using the following hack.

Stick your javascript code in a file .tridactyl.js, for example:

    tri.excmds.sanitise("tridactyllocal tridactylsync");
    tri.excmds.set("searchengine", "duckduckgo");
    tri.excmds.set("hintfiltermode", "vimperator-reflow");
    tri.excmds.set("hintchars", "0123456789");
    tri.scrollone = function() {
        tri.excmds.scrollline(1)
    };
    tri.excmds.bind("e", "jsb tri.scrollone()");

Then in a shell,

    echo 'jsb eval(atob("'`base64 .tridactyl.js`'"))' > .tridactylrc

You could also consider combining this with babel (though to the best of my knowledge babel does not yet support pipelining |> as suggested above).

My example shows what I think is the most useful thing you can do with this, namely, define new functions that you can use in exmode. However to become really useful you probably need to expose more to jsb.

**Comment by bovine3dom** [comment]
Since Tridactyl will be installed on almost everyone's machines by the time we get around to doing v2, it would be nice if we let people write their own excmds more easily (js -p is quite well hidden), but also if we provided an API for other extensions to make their own excmds. I think we should probably namespace them under e.g `contrib.` so that they can't pollute our commands maliciously. Obviously, people could make aliases if they wanted.

**Comment by glacambre** [comment]
>it would be nice if we let people write their own excmds more easily (js -p is quite well hidden)

The more I think about this the more I believe we should get rid of the macro system, the custom compilation pass, typedoc and decorators. I see this as the only way to provide users with the same amount of power we have regarding excmds/completion creation while not having to maintain redundant code.
Related issues: https://github.com/tridactyl/tridactyl/issues/646 , https://github.com/tridactyl/tridactyl/issues/1241

**Comment by cmcaine** [comment]
Yeah, I've thought about tridactyl v2 as a kind of browser lisp machine
before, with most of it configurable from runtime

**Comment by bovine3dom** [comment]
It's probably worth thinking about supporting and using `<all_frames>` in the manifest, not least so that we can have it such that `<Esc>` always brings focus back to the main document.

**Comment by glacambre** [comment]
Related issue: https://github.com/tridactyl/tridactyl/issues/412
